lib/addons/argocd/application.ts

```diff
@@ -1,25 +1,34 @@
 import * as dot from 'dot-object';
-import { GitOpsApplicationDeployment, GitRepositoryReference } from '../../spi';
+import { GitOpsApplicationDeployment, GitRepositoryReference, Values } from '../../spi';
+import { escapeDots } from '../../utils';
 
 /**
  * Argo Application is a utility class that can generate an ArgoCD application
- * from generic GitOps application properties.  
+ * from generic GitOps application properties.
  */
 export class ArgoApplication {
 
-    constructor(private readonly bootstrapRepo: GitRepositoryReference | undefined) {}
+    constructor(private readonly bootstrapRepo: GitRepositoryReference | undefined) { }
 
     public generate(deployment: GitOpsApplicationDeployment, syncOrder?: number) {
 
-        const flatValues = dot.dot(deployment.values);
+        const normalizedValues = this.normalizeValues(deployment.values);
+
+        const flatValues = dot.dot(normalizedValues);
+
         const nameValues = [];
 
-        for( let key in flatValues) {
-            nameValues.push({ name: key, value: `${flatValues[key]}`});
+        for (let key in flatValues) {
+            // Avoid passing the undefined values and empty objects
+            if (flatValues[key] !== undefined) {
```

Comment:
This if statement is a little confusing. You do the same action in both cases, why are you using `else if`? Why not just `if(<all conditions>)`?

Also if you are checking for undefined, what is the purpose of testing for `Object.getPrototypeOf`? Why not just `flatValues[key] === undefined` (or `!==  undefined`? Are you checking for empty arrays?

Reply:
Yes, I am filtering out `undefined` and `empty objects`.

After flatten using `dot.dot(deployment.values)`, we will have in `flatValues`:

```
  'appmeshController.serviceAccount.create': false,
  'appmeshController.serviceAccount.name': 'appmesh-controller',
  'appmeshController.tracing.enabled': false,
  'appmeshController.tracing.provider': 'x-ray',
  'appmeshController.tracing.address': undefined,   <<=== This will be filtered out
  'appmeshController.tracing.port': undefined,   <<=== This will be filtered out
  'appmeshController.enable': true,
  'something': {}   <<===  This will be filtered out, although I cannot recall how this was generated in the first place..
```

The logic of `If else` is:

1. If value is not defined, skip.
2. otherwise:
    - If the value is NOT an object, include it
    - If the value is an object, and it is not empty, include it as well

Reply:
I added a comment to the code.